flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/source/SourceOutputWithWatermarks.java

```diff
@@ -0,0 +1,175 @@
+
+	private final WatermarkOutput onEventWatermarkOutput;
+
+	private final WatermarkOutput periodicWatermarkOutput;
+
+	/**
+	 * Creates a new SourceOutputWithWatermarks that emits records to the given DataOutput
+	 * and watermarks to the (possibly different) WatermarkOutput.
+	 */
+	protected SourceOutputWithWatermarks(
+			PushingAsyncDataInput.DataOutput<T> recordsOutput,
+			WatermarkOutput onEventWatermarkOutput,
+			WatermarkOutput periodicWatermarkOutput,
+			TimestampAssigner<T> timestampAssigner,
+			WatermarkGenerator<T> watermarkGenerator) {
+
+		this.recordsOutput = checkNotNull(recordsOutput);
+		this.onEventWatermarkOutput = checkNotNull(onEventWatermarkOutput);
+		this.periodicWatermarkOutput = checkNotNull(periodicWatermarkOutput);
+		this.timestampAssigner = checkNotNull(timestampAssigner);
+		this.watermarkGenerator = checkNotNull(watermarkGenerator);
+	}
+
+	// ------------------------------------------------------------------------
+	// SourceOutput Methods
+	//
+	// Note that the two methods below are final, as a partial enforcement
+	// of the performance design goal mentioned in the class-level comment.
+	// ------------------------------------------------------------------------
+
+	@Override
+	public final void collect(T record) {
+		collect(record, Long.MIN_VALUE);
+	}
+
+	@Override
+	public final void collect(T record, long timestamp) {
+		try {
+			final long assignedTimestamp = timestampAssigner.extractTimestamp(record, timestamp);
+
+			// IMPORTANT: The event must be emitted before the watermark generator is called.
+			recordsOutput.emitRecord(new StreamRecord<>(record, assignedTimestamp));
+			watermarkGenerator.onEvent(record, assignedTimestamp, onEventWatermarkOutput);
+		} catch (ExceptionInChainedOperatorException e) {
+			throw e;
+		} catch (Exception e) {
+			throw new ExceptionInChainedOperatorException(e);
+		}
+	}
+
+	// ------------------------------------------------------------------------
+	// WatermarkOutput Methods
+	//
+	// These two methods are final as well, to enforce the contract that the
+	// watermarks from emitWatermark(Watermark) go to the same output as the
+	// watermarks from the watermarkGenerator.onEvent(...) calls in the collect(...)
+	// methods.
+	// ------------------------------------------------------------------------
+
+	@Override
+	public final void emitWatermark(Watermark watermark) {
+		onEventWatermarkOutput.emitWatermark(watermark);
+	}
+
+	@Override
+	public final void markIdle() {
+		onEventWatermarkOutput.markIdle();
+	}
+
+	public final void emitPeriodicWatermark() {
+		watermarkGenerator.onPeriodicEmit(periodicWatermarkOutput);
+	}
+
+	// ------------------------------------------------------------------------
+	// Factories
+	// ------------------------------------------------------------------------
+
+	/**
+	 * Creates a new SourceOutputWithWatermarks that emits records to the given DataOutput
+	 * and watermarks to the (possibly different) WatermarkOutput.
```

Comment:
Why is is a possibly different `WatermarkOutput`?

Reply:
For the per-partition watermarking. The records flow directly downstream, but watermarks go to the `WatermarkOutputMultiplexer` which merges them.